### PR TITLE
builtin: inet6_aton function

### PR DESCRIPTION
**Pull request by qhsong**
add `inet6_aton ` builtin function

### CODE REVIEW

File: plan/typeinferer.go [hunk 1]
```diff
@@ -363,7 +363,6 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 	case ast.ConnectionID, ast.InetAton:
 		tp = types.NewFieldType(mysql.TypeLonglong)
 		tp.Flag |= mysql.UnsignedFlag
-	// time related
```
Comment (hanfei1991) [inline plan/typeinferer.go]: Why remove this?

File: plan/typeinferer.go [hunk 2]
```diff
@@ -374,9 +373,6 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 	case ast.Now, ast.Sysdate, ast.CurrentTimestamp, ast.UTCTimestamp:
 		tp = types.NewFieldType(mysql.TypeDatetime)
 		tp.Decimal = v.getFsp(x)
-	// string related
```
Comment (hanfei1991) [inline plan/typeinferer.go]: and this comment?

File: expression/builtin_miscellaneous.go [hunk 3]
```diff
@@ -326,7 +327,58 @@ type builtinInet6AtonSig struct {
...
+	}
+
+	var result []byte
+	if isMappedIpv6 || ip.To4() == nil {
+		result = make([]byte, net.IPv6len)
+	} else {
+		result = make([]byte, net.IPv4len)
+	}
+
+	for i := range result {
```
Comment (hanfei1991) [inline expression/builtin_miscellaneous.go]: This is not necessary.

File: plan/typeinferer.go (hunk 1 above)
Comment (qhsong) [inline plan/typeinferer.go]: the master branch have out of date when I submit pr and this file have change a lot, so I merge some modify and rebase to master cause this problem.I think this is useless.

File: plan/typeinferer.go (hunk 2 above)
Comment (qhsong) [inline plan/typeinferer.go]: As previous explain

**Comment by qhsong** [comment]
@hanfei1991 Hi, I have fix this problem, pls review this.

File: expression/builtin_miscellaneous.go [hunk 4]
```diff
@@ -326,7 +327,54 @@ type builtinInet6AtonSig struct {
...
+
+	if len(ipAddress) == 0 {
+		return d, nil
+	}
+
+	ip := net.ParseIP(ipAddress)
+	if ip == nil {
+		return d, nil
+	}
+
```
Comment (tiancaiamao) [inline expression/builtin_miscellaneous.go]: I don't understand the following code...

why not just return here directly?

```
d.SetBytes([]byte(ip))
return d, nil
```

Comment (qhsong) [inline expression/builtin_miscellaneous.go]: we can not directly return cus golang treat mapped IPv6 address as an IPv4 address, see here:
https://github.com/golang/go/blob/master/src/net/ip_test.go#L22

But mysql treat it as IPv6 address:


So, we need some code to deal this situation.So mysql need `is_ipv4_mapped` to check this ip address.

**Comment by hanfei1991** [comment]
LGTM

File: expression/builtin_miscellaneous.go (hunk 4 above)
Comment (tiancaiamao) [inline expression/builtin_miscellaneous.go]: The result is the same, even IPv4 is alloc with IPv6len, and filled the necessary byte, see here:
https://golang.org/src/net/ip.go#L43

You can test it.

Comment (qhsong) [inline expression/builtin_miscellaneous.go]: @tiancaiamao got it, seems I make a mistakes.If that, the code can be more clean.
I will change it tonight.

Comment (qhsong) [inline expression/builtin_miscellaneous.go]: @tiancaiamao After consider, I think my code is right. For IPv4 , `inet6_aton` will return 8 bytes. Directly return `ip` can not got a correct result.

**Comment by shenli** [comment]
LGTM
@qhsong Thanks！

**Pull request merged**